### PR TITLE
ingester: ActiveSeriesRequest can return native histograms information

### PR DESCRIPTION
**Pull request by krajorama**
#### What this PR does

ingester: ActiveSeriesRequest can return native histograms information

The interface  `ActiveSeries(ActiveSeriesRequest)`, gets a new `Type` field that requests active series by default (0) but can request native histograms (1). In case of histograms the new  bucket_count array is filled with bucket counts in the response. Both these changes should be backwards compatible, enabling a smooth upgrade.

Internally I flipped the order of postings iterators from sharding over active series to filtering active native histogram series over sharding to be able to get the bucket counts from the postings iterator directly. Also removes the use of the intermediate postings list to be able to directly stream the results.

Two alternative were considered:
1) add a new RPC interface. Lot's of duplicate code.
2) encode bucket count in a special label. Due to tag=stringlabels this is not trivial and also would involve extra processing to put the new label into its properly sorted position in the labels. Overall sounds messy.

#### Which issue(s) this PR fixes or relates to

Related to #7981 

#### Checklist

- [X] Tests updated.
- N/A Documentation added.
- N/A deferred `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- N/A [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by flxbk — APPROVED** [review]
Looks good to me. Do you have benchmark results to post by any chance?

**Comment by krajorama** [comment]
```
$ benchstat old.txt new.txt 
goos: linux
goarch: amd64
pkg: github.com/grafana/mimir/pkg/ingester
cpu: AMD Ryzen 7 4700U with Radeon Graphics         
                                       │   old.txt   │              new.txt               │
                                       │   sec/op    │   sec/op     vs base               │
Ingester_ActiveSeries/few_series-8       4.848µ ± 2%   4.761µ ± 1%  -1.77% (p=0.025 n=10)
Ingester_ActiveSeries/~10%_of_series-8   2.220m ± 3%   2.122m ± 2%  -4.40% (p=0.000 n=10)
geomean                                  103.7µ        100.5µ       -3.10%

                                       │   old.txt    │               new.txt               │
                                       │     B/op     │     B/op      vs base               │
Ingester_ActiveSeries/few_series-8       2.461Ki ± 0%   2.406Ki ± 0%  -2.22% (p=0.000 n=10)
Ingester_ActiveSeries/~10%_of_series-8   618.4Ki ± 0%   618.3Ki ± 0%  -0.01% (p=0.000 n=10)
geomean                                  39.01Ki        38.57Ki       -1.12%

                                       │   old.txt   │               new.txt                │
                                       │  allocs/op  │  allocs/op   vs base                 │
Ingester_ActiveSeries/few_series-8        48.00 ± 0%    48.00 ± 0%       ~ (p=1.000 n=10) ¹
Ingester_ActiveSeries/~10%_of_series-8   8.437k ± 0%   8.437k ± 0%       ~ (p=1.000 n=10) ¹
geomean                                   636.4         636.4       +0.00%
¹ all samples are equal
```

**Comment by krajorama** [comment]
```
$ benchstat old.txt new.txt 
goos: linux
goarch: amd64
pkg: github.com/grafana/mimir/pkg/ingester
cpu: AMD Ryzen 7 4700U with Radeon Graphics         
                                       │   old.txt    │               new.txt               │
                                       │    sec/op    │   sec/op     vs base                │
Ingester_ActiveSeries/few_series-8       4.824µ ±  6%   4.727µ ± 1%   -2.01% (p=0.007 n=10)
Ingester_ActiveSeries/~10%_of_series-8   2.506m ± 12%   2.222m ± 1%  -11.36% (p=0.005 n=10)
geomean                                  110.0µ         102.5µ        -6.80%

                                       │   old.txt    │               new.txt               │
                                       │     B/op     │     B/op      vs base               │
Ingester_ActiveSeries/few_series-8       2.461Ki ± 0%   2.430Ki ± 0%  -1.27% (p=0.000 n=10)
Ingester_ActiveSeries/~10%_of_series-8   618.4Ki ± 0%   618.3Ki ± 0%  -0.01% (p=0.000 n=10)
geomean                                  39.01Ki        38.76Ki       -0.64%

                                       │   old.txt   │               new.txt                │
                                       │  allocs/op  │  allocs/op   vs base                 │
Ingester_ActiveSeries/few_series-8        48.00 ± 0%    48.00 ± 0%       ~ (p=1.000 n=10) ¹
Ingester_ActiveSeries/~10%_of_series-8   8.437k ± 0%   8.437k ± 0%       ~ (p=1.000 n=10) ¹
geomean                                   636.4         636.4       +0.00%
¹ all samples are equal
```

File: pkg/ingester/activeseries/active_native_histogram_postings.go [hunk 1]
```diff
@@ -7,6 +7,13 @@ import (
 	"github.com/prometheus/prometheus/tsdb/index"
 )
 
+type BucketCountPostings interface {
+	index.Postings
+	// NativeHistogramBuckets returns the number of buckets for the series reference.
+	// If the series reference is not active, it returns false.
```
Comment (flxbk) [inline pkg/ingester/activeseries/active_native_histogram_postings.go:13]: This does not apply any more to the current implementation I think.

File: pkg/ingester/active_series_test.go [hunk 2]
```diff
@@ -65,21 +67,92 @@ func TestIngester_ActiveSeries(t *testing.T) {
 	returnedSeriesCount := 0
 	for _, res := range server.responses {
 		returnedSeriesCount += len(res.Metric)
+		// Check that all series has the expected number of labels.
```
Comment (flxbk) [inline pkg/ingester/active_series_test.go:70]:
```suggestion
		// Check that all series have the expected number of labels.
```

**Review by flxbk — COMMENTED** [review]
Looks good to me with some small nits.

**Pull request merged**